metrics/compile_meta_collector.go

```diff
@@ -0,0 +1,175 @@
+
+const compileMetaMetricPrefix = "syn_lieutenant_cluster_compile_meta"
+
+var metaLastCompileDesc = prometheus.NewDesc(
+	compileMetaMetricPrefix+"_last_compile",
+	"The timestamp of the last cluster compilation as unix timestamp in seconds.",
+	[]string{"cluster", "tenant"},
+	nil,
+)
+
+// commodore build info has dynamic labels
+func newMetaCommodoreBuildInfoDesc(lbls ...string) *prometheus.Desc {
+	return prometheus.NewDesc(
+		compileMetaMetricPrefix+"_commodore_build_info",
+		"Commodore reported build information.",
+		lbls,
+		nil,
+	)
+}
+
+var metaGlobalDesc = prometheus.NewDesc(
+	compileMetaMetricPrefix+"_global",
+	"Version information of the global defaults commodore repository.",
+	[]string{"cluster", "tenant", "url", "path", "version", "gitSha"},
+	nil,
+)
+
+var metaTenantDesc = prometheus.NewDesc(
+	compileMetaMetricPrefix+"_tenant",
+	"Version information of the tenant commodore repository.",
+	[]string{"cluster", "tenant", "url", "path", "version", "gitSha"},
+	nil,
+)
+
+var metaPackageDesc = prometheus.NewDesc(
+	compileMetaMetricPrefix+"_package",
+	"Version information of the used commodore package repositories.",
+	[]string{"cluster", "tenant", "name", "url", "path", "version", "gitSha"},
+	nil,
+)
+
+var metaInstanceDesc = prometheus.NewDesc(
+	compileMetaMetricPrefix+"_instance",
+	"Version information of the used commodore instance (component) repositories.",
+	[]string{"cluster", "tenant", "name", "url", "path", "version", "gitSha", "component"},
+	nil,
+)
+
+// CompileMetaCollector is a Prometheus collector that translates the `status.compileMeta` field of the Cluster CRD into Prometheus metrics.
+type CompileMetaCollector struct {
+	Client client.Client
+
+	Namespace string
+}
+
+var _ prometheus.Collector = &CompileMetaCollector{}
+
+// Describe implements prometheus.Collector.
+// This collector does not send any descriptions and thus makes the collector unchecked.
+// An unchecked collector is needed because the commodore build info metric is dynamic and has no static description.
+func (*CompileMetaCollector) Describe(_ chan<- *prometheus.Desc) {}
+
+// Collect implements prometheus.Collector.
+// Iterates over all clusters and sends compileMeta metrics for each cluster.
+func (m *CompileMetaCollector) Collect(ch chan<- prometheus.Metric) {
+	ctx := context.Background()
+
+	cls := synv1alpha1.ClusterList{}
+	if err := m.Client.List(ctx, &cls, client.InNamespace(m.Namespace)); err != nil {
+		err := fmt.Errorf("failed to list clusters: %w", err)
+		ch <- prometheus.NewInvalidMetric(metaLastCompileDesc, err)
+		ch <- prometheus.NewInvalidMetric(newMetaCommodoreBuildInfoDesc(), err)
+		ch <- prometheus.NewInvalidMetric(metaGlobalDesc, err)
+		ch <- prometheus.NewInvalidMetric(metaTenantDesc, err)
+		ch <- prometheus.NewInvalidMetric(metaPackageDesc, err)
+		ch <- prometheus.NewInvalidMetric(metaInstanceDesc, err)
+	}
+
+	for _, cl := range cls.Items {
+		ch <- prometheus.MustNewConstMetric(
```

Comment:
How tricky would it be to only generate the last compile, build info, global and tenant repo metrics when the cluster actually has compile meta information?

Reply:
Not that hard.

I thought about that but decided that it feels the most natural to me to leave them empty for a query to find clusters without metadata yet. It should not be that many metrics overall with all the component instance and package metrics.